app/models/miq_server/worker_management/monitor.rb

```diff
@@ -84,6 +78,22 @@ def cleanup_orphaned_worker_rows
     end
   end
 
+  def monitor_active_workers
+    # When k8s or systemd is operating as the worker monitor then all of the
+    # worker monitoring (liveness, memory threshold) is handled by those
+    # systems.  Only when workers are run as standalone processes does MiqServer
+    # have to monitor the workers itself.
+    return if podified? || systemd?
```

Comment:
So looking at this method it
1. Updates the `MiqWorker#last_heartbeat` from the `tmp/guid.hb` heartbeat files
2. Validates the worker heartbeat and memory limits [[ref]](https://github.com/ManageIQ/manageiq/blob/master/app/models/miq_server/worker_management/monitor/validation.rb#L4-L20)

After this PR neither of these will be necessary neither on k8s (no access to heartbeat files and mem/cpu limits handled by k8s) nor on systemd (systemd-notify for heartbeating and cgroups for mem/cpu limits)

Reply:
Side note, in a lot of places we ask `if podified?` and `if systemd?` but we have no way of asking if we're using the third monitor type, process.  Ideally here I'd like to say `return unless worker_monitor_type_process?` since we have three worker monitor types: k8s, systemd, and legacy/process.

Reply:
Side note 2, I am seeing a few categories of "worker_monitor_types" that we make decisions on:
1) external_liveness_monitor?
2) external_readiness_monitor? 
3) external_resource_limit_monitor? 
4) isolated processes:  decisions about where a piece of work can run:  can't get process information for workers from the server or vice versa in pods.

I'm sure there are others but maybe we can come up with good names for these concerns.

Reply:
This is a good point, I've been assuming that any "external worker monitor" aka k8s or systemd will take care of all or most of the "every 15 seconds make sure the workers are okay" for us but some may provide most but not all.  I'm thinking when we move to a more pluggable worker monitor class these won't be questions that have to be asked but rather methods that are implemented or not and should clean things up nicely (:crossed_fingers:)

Unsure of readiness vs liveness, is readiness the "are there enough resources to start this new worker" check?

Can you expand on "decisions about where a piece of work can run"?  Is that "make MiqSever more like k8s and schedule workers across the appliances"?  If so yeah that is a big one that I haven't even started to tackle yet, I think this is going to be a fundamental rethink of how we schedule workers (per server vs per region).  We "got away with it" on podified because we went to a single `MiqServer` but with actual appliance VMs maybe the worker counts should be at the region level and divvied up by the master MiqServer?

Reply:
> I'm thinking when we move to a more pluggable worker monitor class these won't be questions that have to be asked but rather methods that are implemented or not and should clean things up nicely

👍 

> Unsure of readiness vs liveness, is readiness the "are there enough resources to start this new worker" check?

liveness == heartbeat check
readiness == rails server is available (port 3000) for puma based workers


> Can you expand on "decisions about where a piece of work can run"?

There are places that have expectations that the server and workers reside on the same process space / filesystem.  For example:  https://github.com/ManageIQ/manageiq/issues/20835


The workaround for these has been to make the right process fetch the environmental information for itself instead of doing it for someone else.  Generally, it's been places where the assumption was that we needed to run on the miq server.  Previously, any worker on that server was fine but doesn't work in pods so it makes sense to make everything queue things for the correct process instead of assuming one process can see other processes.  Like this:  https://github.com/jrafanie/manageiq/commit/1c3d8a7dc3ffa349daf0d268120eeca6a42487a8

Another example: https://github.com/ManageIQ/manageiq/pull/20290

Reply:
> There are places that have expectations that the server and workers reside on the same process space / filesystem. For example

Ahh yes okay I see what you mean, basically anything doing "worker management" outside of MiqServer based on the assumption it can do it based on the PID (processInfo(pid) in the first example and kill(pid) in the second)

Reply:
Exactly, and i believe we had a similar change for embedded ansible since that relied on the filesystem having the ansible repo checked out.  It might have changed since but the idea was that you can't assume the "server with embedded ansible" can access the locally checked out ansible repo, only the process that's guaranteed to have checked it out.  I don't know what to call this but the process isolation of process space (kill/processInfo(pid)) and filesystems (ansible repo as an example) are what I'm describing.